### PR TITLE
Add related.entity to cloudtrail integration

### PR DESCRIPTION
**Pull request by romulets**
**tldr; Adding `related.entity` to cloudtrail is part of the initial steps of CDR ([epic overview](https://github.com/elastic/security-team/issues/10152)). That enables our customers to better correlate events based on entities and enhances the investigation workflows for the Cloud Security use case.**

[ECS `related.entity` PR](https://github.com/elastic/ecs/pull/2360)

### Background

Elastic Cloud Security Team has been focusing, this past year, on Cloud Detection and Response (CDR). One of the first steps towards the CDR vision is [to enhance investigation workflows for the Cloud Security use-case in SIEM](https://github.com/elastic/security-team/issues/9979).

As part of enhancing investigation workflows it's necessary to be able to correlate events and entities. Meaning, if an alert is triggered on the ec2 instance `i-000000000`, it is of great value to easily be able to search all the events related to that entity, across multiple indices, with one query. [Therefore we are working on extracting entities and enabling them to be correlated](https://github.com/elastic/security-team/issues/9426).

### Why `related.entity`

With this background, we've researched a few options on what would be the best approach to enable such feature (discussions [https://github.com/elastic/security-team/issues/10026](https://github.com/elastic/security-team/issues/10026) and [https://github.com/elastic/security-team/issues/9798](https://github.com/elastic/security-team/issues/9798), and outcomes [https://github.com/elastic/security-team/issues/10152](https://github.com/elastic/security-team/issues/10152)), and arrived at the ecs field `related`.

Based on the `related` description:

> This field set is meant to facilitate pivoting around a piece of data.
> 
> Some pieces of information can be seen in many places in an ECS event.  
> To facilitate searching for them, store an array of all seen values to their  
> corresponding field in `related.`.

To add a broad `related.entity` field that can hold any needed identifier to pivot data on seems to be well fitted. This would enable customers to simply run `related.entity: "i-000000000"` and get all the hits to that specific cloud resource.

### What is an `entity`?

An "entity" in our context refers to any discrete component within an IT environment that can be uniquely identified and monitored. This broad term encompasses both managed and unmanaged elements.

The term "entity" is broader than the current set of available fields under `related`. Although `ip`, `user` and `hosts` can be identities, there is a lack of space to represent messaging queues, load balancers, storage systems, databases and others. Therefore the proposal to add a new field.

### Related issues

- Closes https://github.com/elastic/security-team/issues/9867

### CODE REVIEW

File: packages/aws/changelog.yml [hunk 1]
```diff
@@ -1,4 +1,9 @@
 # newer versions go on top
+- version: "2.25.0-preview01"
+  changes:
+    - description: Add related.entity field.
+      type: enhancement
+      link: https://github.com/elastic/integrations/pull/10216
```
Comment (andrewkroh) [inline packages/aws/changelog.yml:6]: Wrong PR number.

```suggestion
      link: https://github.com/elastic/integrations/pull/11115
```

**Review by andrewkroh — COMMENTED** [review]
Can you please add a link to the ECS `related.entity` PR in the description.

Can you post before/after results of the benchmark for this pipeline (see `elastic-package benchmark pipeline -d cloudtrail -v`, although I'm not sure how stable / repeatable these are at such low event counts).

**Comment by andrewkroh** [comment]
@romulets Can you please perform one more test where you inline the processors into `default.yml`. See my updated comment in https://github.com/elastic/integrations/pull/11115#issuecomment-2359341532. Possibly that combined with your usage of `script` could yield even better results.

---

> Obs: Right now, I'm sorting the entities list only for tests purposes, there is no need to do so on production data. That consumes a some processing and it's unfortunate to have it for testing purposes only.

We could propose a change to `elastic-package` to handle this. How much does the sorting cost?

**Comment by romulets** [comment]
@andrewkroh Indeed, the gains are relevant. I committed the changes.

Updated benchmarks:

Code | main | PR | Three brackets | Painless Script | Inline in default.yml
-- | -- | -- | -- | -- | --
Commit | [076cfec0](https://github.com/romulets/integrations/tree/076cfec0fe501f91d3ba3ae23920f08a5bb0e08b) | [3f5ac7f0](https://github.com/romulets/integrations/tree/3f5ac7f0517fb7d6ef4e091603586e986b26172e) | [36baf08b](https://github.com/romulets/integrations/tree/36baf08b837e3ac219366d1277654ef777f3dc8a) | [8d85643d](https://github.com/romulets/integrations/tree/8d85643dae5280b9e1eb85e34f6fe87632847f7a) | [3f819532](https://github.com/romulets/integrations/tree/3f8195320dda65dc7542e85a0dc789e422e83211)
Average EPS | 16209.72785 | 10723.17927 | 10930.95625 | 13152.65734 | 15102.77816
Stddev | 953.7813683 | 565.3165502 | 524.9144185 | 879.7901167 | 772.4165627
Min | 11111.11111 | 9090.90909 | 9345.79439 | 9708.73786 | 12195.12195
Max | 17543.85965 | 11494.25287 | 11494.25287 | 14285.71429 | 16129.03226
EPS AVG Change to baseline |   | -33.85% | -32.57% | -18.86% | -6.83%
EPS Max Change to baseline |   | -34.48% | -34.48% | -18.57% | -8.06%

**Review by andrewkroh — APPROVED** [review]
LGTM. Thanks for going to the performance testing to get to an optimized version.

File: packages/aws/data_stream/cloudtrail/fields/fields.yml [hunk 2]
```diff
@@ -187,3 +187,5 @@
           type: flattened
           description: >-
             Additional insight details.
+- name: related.entity
+  type: keyword
```
[the remark this answers is not in the document]
Comment (romulets) [reply packages/aws/data_stream/cloudtrail/fields/fields.yml:196]: I haved added a description. Thanks for through review!

File: packages/aws/data_stream/cloudtrail/elasticsearch/ingest_pipeline/default.yml [hunk 3]
```diff
@@ -47,6 +47,167 @@ processors:
...
+      lang: painless
+      ignore_failure: true
+      on_failure: 
+        - set:
+            description: Add error reason
+            field: error.message
+            value: "{{{ _ingest.on_failure_message }}}"
+      source: |
+        void addEntity(Set entities, String fieldName) {
+          entities.add(field(fieldName).get(null));
```
Comment (devamanv) [inline packages/aws/data_stream/cloudtrail/elasticsearch/ingest_pipeline/default.yml:62]: the script adds null values explicitly. Instead of adding null and then removing it later as in line 205-206, should we consider adding a check here to prevent null from being added in the first place?

File: packages/aws/data_stream/cloudtrail/elasticsearch/ingest_pipeline/default.yml [hunk 4]
```diff
@@ -47,6 +47,167 @@ processors:
...
+
+        Set entities = new HashSet();
+
+        addEntity(entities, "json.userIdentity.arn");
+        addEntity(entities, "json.userIdentity.identityProvider");
+        addEntity(entities, "json.userIdentity.principalId");
+        addEntity(entities, "json.userIdentity.sessionContext.sessionIssuer.arn");
+        addEntity(entities, "json.userIdentity.sessionContext.sessionIssuer.userName");
+        addEntity(entities, "json.userIdentity.sessionContext.webIdFederationData.federatedProvider");
+        addEntity(entities, "json.userIdentity.userName");
```
Comment (devamanv) [inline packages/aws/data_stream/cloudtrail/elasticsearch/ingest_pipeline/default.yml:73]: For better readability, IMO we should group related fields and use a `for` loop instead of  repeated calls to `addEntity`, like so:
```
String[] awsUserIdentityFields = new String[] {
  "json.userIdentity.arn",
  "json.userIdentity.identityProvider",
  "json.userIdentity.principalId",
  "json.userIdentity.sessionContext.sessionIssuer.arn",
  "json.userIdentity.sessionContext.sessionIssuer.userName",
  "json.userIdentity.sessionContext.webIdFederationData.federatedProvider",
  "json.userIdentity.userName"
};
for (String field : awsUserIdentityFields) {
  addEntity(entities, field);
}
```
If we happen to do this, consider making the change for other fields too.

File: packages/aws/docs/cloudtrail.md [hunk 5]
```diff
@@ -133,6 +133,7 @@ Please refer to the following [document](https://www.elastic.co/guide/en/ecs/cur
 | host.os.codename | OS codename, if any. | keyword |
 | input.type | Input type | keyword |
 | log.offset | Log offset | long |
+| related.entity | All the entity identifiers related to the document. If the document  contains multiple entities, identifiers belonging to different entities  will be present. Example identifiers include cloud resource IDs, ARNs,  email addresses, or hostnames. | keyword |
```
Comment (devamanv) [inline packages/aws/docs/cloudtrail.md:136]:
```suggestion
| related.entity | A collection of all entity identifiers associated with the document. If the document  contains multiple entities, identifiers for each will be included. Example identifiers include(but not limited to) cloud resource IDs, ARNs,  email addresses, and hostnames. | keyword |
```

File: packages/aws/data_stream/cloudtrail/elasticsearch/ingest_pipeline/default.yml (hunk 3 above)
Comment (romulets) [reply packages/aws/data_stream/cloudtrail/elasticsearch/ingest_pipeline/default.yml:62]: Tackling it!

File: packages/aws/data_stream/cloudtrail/elasticsearch/ingest_pipeline/default.yml (hunk 4 above)
Comment (romulets) [reply packages/aws/data_stream/cloudtrail/elasticsearch/ingest_pipeline/default.yml:73]: hm, I'm not 100% onboard on making more syntax sugar on this. The readability isn't bad right now, all the fields are aligned and easy to read. What we loose is that we have more loops and nesting to read.

It's all taste based. Anyway, I took the time and did. Let me know if you are more satisfied with the readability.

File: packages/aws/docs/cloudtrail.md (hunk 5 above)
Comment (romulets) [reply packages/aws/docs/cloudtrail.md]: tackled!

File: packages/aws/data_stream/cloudtrail/elasticsearch/ingest_pipeline/default.yml [hunk 6]
```diff
@@ -47,6 +47,209 @@ processors:
...
         drop(ctx);
+  # Before renames because some json fields are mutated on the process
+  - script:
+      description: Appends any relevant entity to `related.entity` for all events
+      lang: painless
+      on_failure: 
+        - set:
+            description: Add error reason
+            field: error.message
+            value: "{{{ _ingest.on_failure_message}}}"
```
Comment (devamanv) [inline packages/aws/data_stream/cloudtrail/elasticsearch/ingest_pipeline/default.yml:58]:
```suggestion
            value: "{{{ _ingest.on_failure_message }}}"
```

File: packages/aws/data_stream/cloudtrail/elasticsearch/ingest_pipeline/default.yml [hunk 7]
```diff
@@ -47,6 +47,209 @@ processors:
...
+      source: |
+        void addFields(Set entities, String[] fields) {
+            for (String field : fields) {
+              addField(entities, field);
+            }
+        }
+
+        void addField(Set entities, String fieldName) {
```
Comment (devamanv) [inline packages/aws/data_stream/cloudtrail/elasticsearch/ingest_pipeline/default.yml:66]: Unless I'm missing something, we could simply call `addValue` here and move `field(fieldName).get(null)` inside the function instead of creating an extra function `addField`

**Review by devamanv — APPROVED** [review]
Left a couple of comments. Rest looks good to me.

File: packages/aws/data_stream/cloudtrail/elasticsearch/ingest_pipeline/default.yml (hunk 4 above)
Comment (devamanv) [reply packages/aws/data_stream/cloudtrail/elasticsearch/ingest_pipeline/default.yml]: Thanks for taking the time out to make the change. Changes look good, just one [comment](https://github.com/elastic/integrations/pull/11115#discussion_r1777713869) about the helper functions created

File: packages/aws/data_stream/cloudtrail/elasticsearch/ingest_pipeline/default.yml (hunk 7 above)
Comment (romulets) [reply packages/aws/data_stream/cloudtrail/elasticsearch/ingest_pipeline/default.yml:66]: sometimes we add one field only, so that is for that case mostly

**Comment by efd6** [comment]
I think this should have been 2.26.0-preview01.
